app/src/main/java/com/google/samples/apps/sunflower/adapters/PlantAdapter.kt

```diff
@@ -23,37 +23,93 @@ import androidx.navigation.findNavController
 import androidx.recyclerview.widget.DiffUtil
 import androidx.recyclerview.widget.ListAdapter
 import androidx.recyclerview.widget.RecyclerView
+import androidx.recyclerview.widget.StaggeredGridLayoutManager
 import com.google.samples.apps.sunflower.PlantListFragment
-import com.google.samples.apps.sunflower.PlantListFragmentDirections
+import com.google.samples.apps.sunflower.R
+import com.google.samples.apps.sunflower.HomeViewPagerFragmentDirections
 import com.google.samples.apps.sunflower.data.Plant
 import com.google.samples.apps.sunflower.databinding.ListItemPlantBinding
 
 /**
  * Adapter for the [RecyclerView] in [PlantListFragment].
  */
-class PlantAdapter : ListAdapter<Plant, PlantAdapter.ViewHolder>(PlantDiffCallback()) {
+class PlantAdapter : ListAdapter<Plant, RecyclerView.ViewHolder>(PlantDiffCallback()) {
 
-    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
-        val plant = getItem(position)
-        holder.apply {
-            bind(createOnClickListener(plant.plantId), plant)
-            itemView.tag = plant
+    /**
+     * Int array of all the positions with header-style cards
+     */
+    private val headerPositions = intArrayOf(0)
+
+    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
+
+        // if current position is a header element
+        if (position in headerPositions) {
```

Comment:
`when`ing over the itemType seems cleaner/safer.

Reply:
Not sure what you mean - `getItem` always returns a `Plant`, so a `when` would always resolve to `is Plant -> ...` ?

Reply:
If I'm not wrong please correct me @nickbutcher, normally when working with adapters is a good practice to create an enum or constants representing the different cells type. This is the best way to add new cells types to the current list or implement a specific logic for them.

Having in we case the following code structure:
```java
override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (getItemViewType(position)) {
            VIEW_TYPE_HEADER -> ...
            VIEW_TYPE_PLANT -> ...
        }
    }
```
If you want a detailed example please look this one: https://github.com/android/plaid/blob/master/about/src/main/java/io/plaidapp/about/ui/adapter/LibraryAdapter.kt

Reply:
Sorry this wasn't clear. Yes I meant `when(getItemViewType(position)`